build/build.go

```diff
@@ -1612,3 +1619,40 @@ func noPrintFunc(opt map[string]Options) bool {
 	}
 	return true
 }
+
+func saveLocalState(so client.SolveOpt, opt Options, node builder.Node, configDir string) error {
+	var err error
+
+	if so.Ref == "" {
+		return nil
+	}
+
+	lp := opt.Inputs.ContextPath
+	dp := opt.Inputs.DockerfilePath
+	if lp != "" || dp != "" {
+		if lp != "" {
+			lp, err = filepath.Abs(lp)
+			if err != nil {
+				return err
+			}
+		}
+		if dp != "" {
+			dp, err = filepath.Abs(dp)
+			if err != nil {
+				return err
+			}
+		}
+		ls, err := localstate.New(configDir)
+		if err != nil {
+			return err
+		}
+		if err := ls.SaveRef(node.Builder, node.Name, so.Ref, localstate.State{
+			LocalPath:      lp,
```

Comment:
Do we want to also set the command being invoked?:

```suggestion
            Cmd:            strings.Join(os.Args[1:], " "),
			LocalPath:      lp,
```

This could be handy as we can't easily reconstruct from frontend attrs.

I also wonder if we could set the Buildx version.

Reply:
I don't think we should save the raw command. That means someone will try to parse the string again and we don't want that. Also, build can be configured in many other ways and non cmd pkg should not access `os.Args`. If there are more fields that we should capture, we should add them directly in a structured way.

I'm not sure about the version as we don't save the BuildKit version either in history API iirc. But maybe there is an use-case you have in mind.

Reply:
> That means someone will try to parse the string again and we don't want that. Also, build can be configured in many other ways and non cmd pkg should not access `os.Args`.

Ah right. Main idea was to track back the command issued for a build but frontend attrs should be enough actually.

> I'm not sure about the version as we don't save the BuildKit version either in history API iirc. But maybe there is an use-case you have in mind.

I was thinking of a diagnostic behavior so user could track back what client version was used for this. BuildKit version looks better for this use case actually, that's a good point. It would be useful to have it in the history for the same use case.